# Import Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Import Dataset

In [2]:
df = pd.read_csv('../data/raw-data/superstore.csv', encoding='latin1')

In [3]:
df

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9989,9990,CA-2014-110422,1/21/2014,1/23/2014,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,...,33180,South,FUR-FU-10001889,Furniture,Furnishings,Ultra Door Pull Handle,25.2480,3,0.20,4.1028
9990,9991,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.9600,2,0.00,15.6332
9991,9992,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.5760,2,0.20,19.3932
9992,9993,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.6000,4,0.00,13.3200


# About the Data

In [4]:
df.shape

(9994, 21)

In [5]:
df.sample(5)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
6342,6343,CA-2015-146255,3/7/2015,3/10/2015,First Class,EM-14140,Eugene Moren,Home Office,United States,Newark,...,19711,East,TEC-AC-10004469,Technology,Accessories,Microsoft Sculpt Comfort Mouse,119.850,3,0.0,52.7340
4474,4475,CA-2017-121300,9/29/2017,9/29/2017,Same Day,MG-17680,Maureen Gastineau,Home Office,United States,Mentor,...,44060,East,OFF-PA-10002787,Office Supplies,Paper,Xerox 227,15.552,3,0.2,5.4432
4146,4147,CA-2015-136196,11/28/2015,12/4/2015,Standard Class,TP-21415,Tom Prescott,Consumer,United States,Freeport,...,11520,East,FUR-FU-10004017,Furniture,Furnishings,Tenex Contemporary Contur Chairmats for Low an...,322.590,3,0.0,64.5180
8087,8088,CA-2016-106621,9/29/2016,10/1/2016,Second Class,DM-12955,Dario Medina,Corporate,United States,Yuma,...,85364,West,OFF-AR-10002375,Office Supplies,Art,Newell 351,10.496,4,0.2,1.1808
2516,2517,CA-2014-128888,11/15/2014,11/22/2014,Standard Class,PB-19105,Peter Bühler,Consumer,United States,Houston,...,77095,Central,OFF-EN-10003001,Office Supplies,Envelopes,Ames Color-File Green Diamond Border X-ray Mai...,604.656,9,0.2,204.0714


In [6]:
df['Country'].unique()

<StringArray>
['United States']
Length: 1, dtype: str

In [7]:
df[df['Order ID'] == 'CA-2016-152156']

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.96,2,0.0,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.94,3,0.0,219.5820


---

By my observation, research and by doing some operations on dataset I get to know that....
The dataste contains the superstore sales data. Total 9994 records with 21 columns. The data only contain the US sales data or maybe the superstore is only available at US.  
The each row in the data represents one product. If one customer purchesed the 2 products at the same time then it mentioned in the data n 2 columns, saperate entry for each product.  

It contains `Row ID`, `Order ID`, `Order Date`, `Ship Date`, `Ship Mode`,	`Customer ID`, `Customer Name`,	`Segment`, `Country`, `City`, `State`, `Postal Code`, `Region`,	`Product ID`, `Category`, `Sub-Category`, `Product Name`, `Sales`. `Quantity`, `Discount` & `Profit`. These are in total 21 columns.

# Columns descriptions

In [8]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


---

**`superstore-sales`**

`Row ID` - It is just a number representing the row number. We have 9800 rows.  
`Order ID` - It is the order ID, It is the same for those product which are ordered at the same time by the same customer.  
`Order Date` - The date of ordered is in the Order Date column  
`Ship Date` - This column contains the date. The date is of the day on which the product is delivered to customer.  
`Ship Mode` - Ship mode column contains the string values. Which have represents the shiping mode of the product.  
`Customer ID` - Customer ID in alphanumeric code. This is the combination of name and surname initials with a unique code concatinated with '-'  
`Customer Name` - Name of the customer.  
`Segment` - This column contains the user segment. The customers are splited into 3 groups. (**Consumer**: Regular people buying things for themselves. **Corporate**: Big companies buying things for their offices. **Home Office**: People who work from home buying things for their workspace.)  
`Country` - Country name which is only United States.
`City` - It has the city name. Where the customer is live or package is delivered.  
`State` - The state.  
`Postal Code` - The postal code.  
`Region` - The US geographical area is splited into 4 parts (Central, East, South & West)  
`Product ID` - It is the unique id for each product. It is like (category-subcategoey-uniquecode)  
`Category` - It has the category of the product. (ex. Furniture, Offie Supplies, Art, ...etc)  
`Sub-Category` - It has the sub-category of the product. (ex. Bookcases, chair, Table, ...etc)  
`Product Name` - The actual name of the product by the brand.  
`Sales` - It is the price of the product.   
`Quantity` - It is the number of particular product that customer order.  
`Discount` - The discount given (e.g., 0.2 = 20%).  
`Profit` - It is the profit made by the superstore, negative value mean loss.  

# Assessment

## Check by eyes
Take a look at any unusual data in the dataset by eyes

In [9]:
# Lets run sample command to get random rows and study them
df.sample(5)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
9943,9944,CA-2014-143371,12/28/2014,1/3/2015,Standard Class,MD-17350,Maribeth Dona,Consumer,United States,Anaheim,...,92804,West,OFF-SU-10002537,Office Supplies,Supplies,Acme Box Cutter Scissors,51.150,5,0.0,13.2990
5659,5660,CA-2016-145261,12/18/2016,12/21/2016,First Class,AH-10120,Adrian Hane,Home Office,United States,Salem,...,97301,West,OFF-LA-10000407,Office Supplies,Labels,Avery White Multi-Purpose Labels,15.936,4,0.2,5.1792
1950,1951,CA-2017-157987,9/2/2017,9/6/2017,Standard Class,AC-10615,Ann Chong,Corporate,United States,New York City,...,10009,East,OFF-BI-10004970,Office Supplies,Binders,"ACCOHIDE 3-Ring Binder, Blue, 1""",26.432,8,0.2,8.9208
3819,3820,CA-2017-141873,10/9/2017,10/14/2017,Standard Class,HG-14845,Harry Greene,Consumer,United States,New York City,...,10024,East,OFF-BI-10001031,Office Supplies,Binders,Pressboard Data Binders by Wilson Jones,12.816,3,0.2,4.3254
5842,5843,CA-2015-126725,11/17/2015,11/21/2015,Standard Class,BS-11665,Brian Stugart,Consumer,United States,San Diego,...,92105,West,OFF-ST-10001418,Office Supplies,Storage,Carina Media Storage Towers in Natural & Black,304.900,5,0.0,6.0980


### Problems found

The column names should be changed, remove the spaces.

---

## Check by info()
Lets check data types of all columns

In [13]:
# info to chack data types of the columns
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   str    
 2   Order Date     9994 non-null   str    
 3   Ship Date      9994 non-null   str    
 4   Ship Mode      9994 non-null   str    
 5   Customer ID    9994 non-null   str    
 6   Customer Name  9994 non-null   str    
 7   Segment        9994 non-null   str    
 8   Country        9994 non-null   str    
 9   City           9994 non-null   str    
 10  State          9994 non-null   str    
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   str    
 13  Product ID     9994 non-null   str    
 14  Category       9994 non-null   str    
 15  Sub-Category   9994 non-null   str    
 16  Product Name   9994 non-null   str    
 17  Sales          9994 non-null   float64
 18  Quantity       9994

In [14]:
# Check for Categorical columns 
df['Ship Mode'].unique(), df['Segment'].unique(), df['Region'].unique(), df['Category'].unique(), df['Sub-Category'].unique()

(<StringArray>
 ['Second Class', 'Standard Class', 'First Class', 'Same Day']
 Length: 4, dtype: str,
 <StringArray>
 ['Consumer', 'Corporate', 'Home Office']
 Length: 3, dtype: str,
 <StringArray>
 ['South', 'West', 'Central', 'East']
 Length: 4, dtype: str,
 <StringArray>
 ['Furniture', 'Office Supplies', 'Technology']
 Length: 3, dtype: str,
 <StringArray>
 [  'Bookcases',      'Chairs',      'Labels',      'Tables',     'Storage',
  'Furnishings',         'Art',      'Phones',     'Binders',  'Appliances',
        'Paper', 'Accessories',   'Envelopes',   'Fasteners',    'Supplies',
     'Machines',     'Copiers']
 Length: 17, dtype: str)

### Problems Found

Here the data type of some columns are not appropriate  
1. `Order Date` and `Ship Date` should be in datetime.
2. `Postal Code` should be str. (Cause some cities have postal code with leading zeros)
3. `Ship Mode`, `Segment`, `Region`, `Category`, `State` & `Sub-Category` should be categorical.

---

## Chech by isnull()
Lets check any null values are present in the dataset

In [15]:
# Lets check null values
df.isnull().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

### Problems Found

No null values in this data

## Check by duplicated()
check for duplicate record in the dataset

In [16]:
# check for duplicate record.
df[df.duplicated()]

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit


In [17]:
# Lets check without Row ID column
df.drop(columns='Row ID', inplace=True)

In [18]:
df[df.duplicated()]

,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
3406,US-2014-150119,4/23/2014,4/27/2014,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,Ohio,43229,East,FUR-CH-10002965,Furniture,Chairs,Global Leather Highback Executive Chair with P...,281.372,2,0.3,-12.0588


In [19]:
# Found one duplicate record 
# Lets first check is it really a duplicate record
df[df['Order ID']=='US-2014-150119']

,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
3405,US-2014-150119,4/23/2014,4/27/2014,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,Ohio,43229,East,FUR-CH-10002965,Furniture,Chairs,Global Leather Highback Executive Chair with P...,281.372,2,0.3,-12.0588
3406,US-2014-150119,4/23/2014,4/27/2014,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,Ohio,43229,East,FUR-CH-10002965,Furniture,Chairs,Global Leather Highback Executive Chair with P...,281.372,2,0.3,-12.0588
3407,US-2014-150119,4/23/2014,4/27/2014,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,Ohio,43229,East,OFF-BI-10000145,Office Supplies,Binders,Zipper Ring Binder Pockets,7.488,8,0.7,-5.2416
3408,US-2014-150119,4/23/2014,4/27/2014,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,Ohio,43229,East,FUR-FU-10002191,Furniture,Furnishings,G.E. Halogen Desk Lamp Bulbs,22.336,4,0.2,7.8176


### Problem Found

Order ID `US-2014-150119` has one duplicate record.

---

## Check by describe()


In [20]:
df.describe()

,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,55190.379428,229.858001,3.789574,0.156203,28.656896
std,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,90008.000000,209.940000,5.000000,0.200000,29.364000
max,99301.000000,22638.480000,14.000000,0.800000,8399.976000


No unusuall data found

---

# Documenting the Assessment

1. Column names should not contain the space. (Remove or replace spaces from the column names).
2. Incorrect Data types:
    1. `Order Date` and `Ship Date` columns have str dtype (Change it to datetime)
    2. `Postal Code` have int64 dtype (Change it to str)
    3. `Ship Mode`, `Segment`, `Region`, `Category`, `State` & `Sub-Category` have dtype str (Change it to Categorical)
3. Order ID `US-2014-150119` has one duplicate record.